### PR TITLE
Issue #1744 Update centos_ci.sh for on demand job for minishift docs

### LINKED ISSUE #1744: Update centos_ci.sh for on demand job for minishift docs
This is aligning with https://github.com/minishift/minishift-ci-jobs/issues/32

### PR DESCRIPTION
**Pull request by budhrg**
Fix #1744

### CODE REVIEW

**Comment by budhrg** [comment]
retest this please

**Comment by budhrg** [comment]
Working as expected http://artifacts.ci.centos.org/minishift/minishift/pr/1745/openshift-origin/latest/minishift/index.html

Need to properly update the PR now.

**Comment by hferentschik** [comment]
What is this about? What's the actual change here? I see there are quite some formatting changes.

File: docs/Rakefile [hunk 1]
```diff
@@ -15,12 +15,12 @@ ADOC_VARIABLES     = FileList.new('source/variables.adoc',
 )
 GENERATED_ADOC_DIR = "source/command-ref"
 ADOC_FILES = Rake::FileList.new do |list|
-	MARKDOWN_FILES.each do |f|
-        adoc = f.ext(".adoc")
-        adoc = File.split(adoc)[1]
-        adoc = File.join(GENERATED_ADOC_DIR, adoc)
-		list.add(adoc)
-	end
+  MARKDOWN_FILES.each do |f|
```
Comment (hferentschik) [inline docs/Rakefile]: Why these changes?

Comment (budhrg) [reply docs/Rakefile]: @hferentschik There are inconsistencies in tabs alignment here.

Like https://github.com/minishift/minishift/blob/master/docs/Rakefile#L139-L143 

and 

https://github.com/minishift/minishift/blob/master/docs/Rakefile#L101-L116

and then this one https://github.com/minishift/minishift/blob/master/docs/Rakefile#L128-L136

These changes automatically got it when I was trying to fix the alignment here https://github.com/minishift/minishift/blob/master/docs/Rakefile#L19-L23.

And since, my Sublime is set to have tab with 2 spaces, other places it got aligned.

Bottom line is we now have tab alignment consistent in `docs/Rakefile` with two space tab.

**Comment by budhrg** [comment]
@hferentschik , this PR change is for the job https://ci.centos.org/job/minishift-docs which I updated to accept two parameters BRANCH and REPO for getting the openshift docs published whenever we want.

So, we need to specify our repo url  and branch (default master) which contains desired docs changes and then this job will take care of publishing the docs like  http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/budhrg-docs/openshift-origin/latest/minishift/index.html.

Publishing docs along with PR (which I did last time) felt like over kill to me as we hardly checking docs that time.

Hene, this on-demand job.

About formatting, I have commented here https://github.com/minishift/minishift/pull/1745#discussion_r152971693

File: centos_ci.sh [hunk 2]
```diff
@@ -292,14 +292,40 @@ function perform_release() {
...
+function perform_docs_publish() {
+  # REPO and BRANCH variables are populated via https://ci.centos.org/job/minishift-docs
+  DOC_USER=$(echo $REPO | cut -d"/" -f4)
+  cd gopath/src/github.com/minishift/minishift
+
+  # Fetch remote repo and checkout to specified branch if not default minishift repo
+  if [[ "$DOC_USER" != "minishift" ]]; then
+    git remote add remote-user $REPO
+    git fetch remote-user
+    git checkout -b remote-docs-branch --track remote-user/$BRANCH
```
Comment (hferentschik) [inline centos_ci.sh]: track this branch?

File: centos_ci.sh [hunk 3]
```diff
@@ -292,14 +292,40 @@ function perform_release() {
   curl http://minibot.19cf262c.svc.dockerapp.io:9009/hubot/centosci -H "Content-Type: application/json" -d '{"payload":{"status":"success","message":'"\"$MESSAGE\""',"url":'"\"$URL\""'}}' || true
 }
 
+function perform_docs_publish() {
+  # REPO and BRANCH variables are populated via https://ci.centos.org/job/minishift-docs
```
Comment (hferentschik) [inline centos_ci.sh]: Do we need to be able to specify the repo?

File: centos_ci.sh [hunk 4]
```diff
@@ -292,14 +292,40 @@ function perform_release() {
...
+  DOC_USER=$(echo $REPO | cut -d"/" -f4)
```
Comment (hferentschik) [inline centos_ci.sh]: DOC_USER seems to be not quite what this variable is about. More like repo owner, right?

File: centos_ci.sh [hunk 5]
```diff
@@ -292,14 +292,40 @@ function perform_release() {
...
+  cd gopath/src/github.com/minishift/minishift
+
+  # Fetch remote repo and checkout to specified branch if not default minishift repo
+  if [[ "$DOC_USER" != "minishift" ]]; then
```
Comment (hferentschik) [inline centos_ci.sh]: This seems odd, should the ability to specify a different repo not be baked in at the top level, instead of just for building the docs?

Comment (hferentschik) [reply centos_ci.sh]: And how is this all used? There seem to be no docs either.

File: docs/Rakefile (hunk 1 above)
Comment (hferentschik) [reply docs/Rakefile]: At least keep this in a different commit, please. I have no problem using the same issue key, but I'd like to distinguish the actual fix for the issue from this unrelated cleanup.

**Comment by hferentschik** [comment]
> Publishing docs along with PR (which I did last time) felt like over kill to me as we hardly checking docs that time.

Sure that part I get.

File: centos_ci.sh (hunk 4 above)
Comment (budhrg) [reply centos_ci.sh]: Ya you are right. Will update it.

File: centos_ci.sh (hunk 3 above)
Comment (budhrg) [reply centos_ci.sh]: All parameters are optional.

So, if I want to test how documentation render for my branch then, I need to specify branch and repo. Otherwise, we need to push our branch to upstream always to test it.

Comment (hferentschik) [reply centos_ci.sh]: But now you can easily get confused about what's getting pushed. One thinks the docs from the master repo are published, but in reality, it is something else.

File: centos_ci.sh [hunk 6]
```diff
@@ -292,14 +292,40 @@ function perform_release() {
...
+    git remote add remote-user $REPO
```
Comment (hferentschik) [inline centos_ci.sh]: remote-user does not seem to be the right name either. Remote repo or custom-repo or custom-remote. You are adding a remote repository not a user.

File: centos_ci.sh [hunk 7]
```diff
@@ -292,14 +292,40 @@ function perform_release() {
...
+
+  install_docs_prerequisite_packages;
+  make gen_adoc_tar
+  build_openshift_origin_docs $(pwd)/docs/build/minishift-adoc.tar;
+
+  mkdir -p minishift/docs/ondemand/$DOC_USER-$BRANCH
+  cp -r openshift-docs/_preview/openshift-origin minishift/docs/ondemand/$DOC_USER-$BRANCH/ # Copy the openshift-origin
+  cp docs/build/minishift-adoc.tar minishift/docs/ondemand/$DOC_USER-$BRANCH/ # Copy doc tar
+  # http://stackoverflow.com/a/22908437/1120530; Using --relative as --rsync-path not working
+  RSYNC_PASSWORD=$RSYNC_PASSWORD rsync -av --delete --relative minishift/docs/ondemand/$DOC_USER-$BRANCH/ minishift@artifacts.ci.centos.org::minishift/
```
Comment (hferentschik) [inline centos_ci.sh]: So where does this publish to? This is not even a single destination, right? This publishes to repo name and branch? Where is this documented?

File: centos_ci.sh (hunk 3 above)
Comment (budhrg) [reply centos_ci.sh]: @hferentschik , I think you are confusing with the intent here.

I have created another issue https://github.com/minishift/minishift/issues/1761 which will maintain a published doc on each PR merge.

The purpose of this PR is to give ability to developer to see how published docs will look like from his desired changes.
This is the path I am assuming user will take when he see any documentation task.

- Create a branch (say `issue-docs-new-cmd`)
- Add necessary documentation changes
- Run locally via `make gen_docs; make serve_docs`.

*NOTE*: Currently, there is no way how you can validate how docs change will look when published after release.

With this PR, developer can see how it will look when published
- User will trigger [minishift-job](https://ci.centos.org/job/minishift-docs/) by passing parameters `BRANCH` ( `issue-docs-new-cmd` in this case) and `REPO` ( `https://github.com/<username>/minishift` in this case).
- Once, developer is sure that the changes looking fine in published doc from above job, he can send PR referring this published doc. It will help reviewer to have and verify the docs changes easily.

Not sure how one can easily get confused as you mentioned in above comments. 

I completely agreed this PR need documentation changes.  

@minishift/minishift-dev  Any thoughts / comments / suggestion.

Comment (hferentschik) [reply centos_ci.sh]: >  I think you are confusing with the intent here.

I understood it once I saw where you are publishing to. 

> Currently, there is no way how you can validate how docs change will look when published after release.

You can build the integrated docs as per documentation, right?

> Not sure how one can easily get confused as you mentioned in above comments.

Since I initially thought this all gets pushed to the same directory. If there would have been some docs, I maybe would have known earlier the intend ;-)

Comment (budhrg) [reply centos_ci.sh]: > You can build the integrated docs as per documentation, right?

Yes, documentation is in place but lot of work. Who else verify the docs change by creating doc tar from someone's PR and serve openshift-docs locally :wink: 

Hence, this PR :) Now, reviewer can just need PR owner repo link and branch and after 10 mins review production like documentation :smile: 

> Since I initially thought this all gets pushed to the same directory. If there would have been some docs, I maybe would have known earlier the intend ;-)

Agreed!  Will take care next time  :+1:

**Comment by robin-owen** [comment]
@budhrg I see that documentation changes have already been mentioned. Do you plan to include the changes in this PR? If so, please add me as a reviewer. :)

**Comment by budhrg** [comment]
> Do you plan to include the changes in this PR? If so, please add me as a reviewer. :)

Sure @robin-owen

File: docs/Rakefile (hunk 1 above)
Comment (budhrg) [reply docs/Rakefile]: Done

File: docs/source/contributing/writing-docs.adoc [hunk 8]
```diff
@@ -159,3 +159,10 @@ $ rake build
 ----
 
 If the build completes successfully, the site is available under *_preview/openshift-origin/latest/welcome/index.html_*.
+
```
Comment (budhrg) [inline docs/source/contributing/writing-docs.adoc]: @robin-owen Please check the documentation here. Let me know your comments/suggestions.

File: docs/source/contributing/writing-docs.adoc [hunk 9]
```diff
@@ -159,3 +159,10 @@ $ rake build
...
+[[building-openshift-docs-automatically]]
+=== Building the OpenShift Documentation Automatically
+
+To automatically get the view of {project} documentation integrated into the OpenShift documentation, trigger the link:https://ci.centos.org/job/minishift-docs/build[minishift docs] job by specifying your repo URL in `REPO` and branch name containing your documentation changes in `BRANCH` parameters respectively. This job will perform all necessary steps including the steps mentioned in xref:manually-building-openshift-docs[Manually Building the OpenShift Documentation].
+
+At the end of the successful job, you will get message where to find the hosted {project} documentation.
```
Comment (robin-owen) [inline docs/source/contributing/writing-docs.adoc]: "receive a message informing you of where to find" or "receive a message with the location of the hosted {project} documentation".

File: docs/source/contributing/writing-docs.adoc [hunk 10]
```diff
@@ -159,3 +159,10 @@ $ rake build
 ----
 
 If the build completes successfully, the site is available under *_preview/openshift-origin/latest/welcome/index.html_*.
+
+[[building-openshift-docs-automatically]]
+=== Building the OpenShift Documentation Automatically
+
+To automatically get the view of {project} documentation integrated into the OpenShift documentation, trigger the link:https://ci.centos.org/job/minishift-docs/build[minishift docs] job by specifying your repo URL in `REPO` and branch name containing your documentation changes in `BRANCH` parameters respectively. This job will perform all necessary steps including the steps mentioned in xref:manually-building-openshift-docs[Manually Building the OpenShift Documentation].
```
Comment (robin-owen) [inline docs/source/contributing/writing-docs.adoc]: One sentence per line, please.

How do I trigger the job? Are the parameters only given through the web UI, or is there a way to provide this information on the command line (or both)? Also, would it be more technically accurate to refer to the "**minishift-docs** project" instead of the "minishift docs job"?

File: docs/source/contributing/writing-docs.adoc [hunk 11]
```diff
@@ -121,8 +121,8 @@ $ make clean_docs
 
 The {project} documentation is deployed on *docs.openshift.org* under link:https://docs.openshift.org/latest/minishift[https://docs.openshift.org/latest/minishift].
 
-To integrate with *docs.openshift.org*, we deliver a tarball that contains the {project} AsciiDoc files as well as some link:http://www.asciibinder.org/[AsciiBinder] metadata files.
-This tarball is generated with a Jenkins job on link:https://ci.centos.org/job/minishift-docs[CentOS CI] and is triggered as part of a xref:./releasing.adoc#cut-release[release].
+To integrate with *docs.openshift.org*, we deliver a tarball at link:http://artifacts.ci.centos.org/minishift/minishift/docs/latest/[artifact server] that contains the {project} AsciiDoc files as well as some link:http://www.asciibinder.org/[AsciiBinder] metadata files.
```
Comment (robin-owen) [inline docs/source/contributing/writing-docs.adoc]: Replace "tarball" with "compressed archive" or similar. Also, I think it would be cleaner to introduce the location of the archive in a separate sentence. Example:

```adoc
To integrate with *docs.openshift.org*, we deliver a compressed archive that contains the {project} AsciiDoc files as well as some link:http://www.asciibinder.org/[AsciiBinder] metadata files.
This archive is generated by a link:https://ci.centos.org/job/minishift-release[release job] which is triggered by an xref:../releasing.adoc#automated-release[automated release].
The generated archive is located on the link:http://artifacts.ci.centos.org/minishift/minishift/docs/latest/[artifact server].
```

File: docs/source/contributing/writing-docs.adoc (hunk 10 above)
Comment (budhrg) [reply docs/source/contributing/writing-docs.adoc]: @robin-owen , only UI.

Also, referring as `minishift-docs job`. There is no project as such.

**Comment by budhrg** [comment]
@hferentschik @robin-owen This is how it is looking like now
screenshot from 2017-11-29 13-16-20

File: docs/source/contributing/writing-docs.adoc [hunk 12]
```diff
@@ -159,3 +160,11 @@ $ rake build
 ----
 
 If the build completes successfully, the site is available under *_preview/openshift-origin/latest/welcome/index.html_*.
+
+[[building-openshift-docs-automatically]]
+=== Building the OpenShift Documentation Automatically
```
Comment (hferentschik) [inline docs/source/contributing/writing-docs.adoc]: "Building the OpenShift Documentation using CentOS CI"

'Automatically' is really ambiguous and not really what this is about

File: docs/source/contributing/writing-docs.adoc [hunk 13]
```diff
@@ -159,3 +160,11 @@ $ rake build
...
+
+To automatically get the view of {project} documentation integrated into the OpenShift documentation, open the link:https://ci.centos.org/job/minishift-docs/[minishift-docs] job in your browser and trigger it by clicking on `Build with Parameters` and specifying your repo URL in `REPO` and branch name containing your documentation changes in `BRANCH` parameters respectively.
```
Comment (hferentschik) [inline docs/source/contributing/writing-docs.adoc]: "To stage a version of the documentation based on a given repository and branch, you can use the CentOS CI  link:https://ci.centos.org/job/minishift-docs/[minishift-docs] job.

Comment (hferentschik) [reply docs/source/contributing/writing-docs.adoc]: I don't see a "Build with Parameters". I would think to do so, one needs to be logged in, but from what I recall from our discussions, we are not supposed to log into the instance. I cannot find the username/password either anyway. We have the access token, but that's a different beast, right? Something is not adding up here.

File: docs/source/contributing/writing-docs.adoc [hunk 14]
```diff
@@ -159,3 +160,11 @@ $ rake build
...
+This job will perform all necessary steps including the steps mentioned in xref:manually-building-openshift-docs[Manually Building the OpenShift Documentation].
+
+At the end of the successful job, you will receive a message with the location of the hosted {project} documentation.
```
Comment (hferentschik) [inline docs/source/contributing/writing-docs.adoc]: Receive a message how? I guess you mean the build log will contain the URL? Can we not display a generic form of the URL as we do in other places?

File: centos_ci.sh [hunk 15]
```diff
@@ -292,14 +292,42 @@ function perform_release() {
   curl http://minibot.19cf262c.svc.dockerapp.io:9009/hubot/centosci -H "Content-Type: application/json" -d '{"payload":{"status":"success","message":'"\"$MESSAGE\""',"url":'"\"$URL\""'}}' || true
 }
 
+function perform_docs_publish() {
+  # REPO and BRANCH variables are populated via https://ci.centos.org/job/minishift-docs
+  REPO_OWNER=$(echo $REPO | cut -d"/" -f4)
+  cd gopath/src/github.com/minishift/minishift
+
+  # Get changes from remote repo/branch
+  if [[ "$REPO_OWNER" != "minishift" ]]; then
```
Comment (hferentschik) [inline centos_ci.sh]: So this if statement is for the case that only the branch is specified? You don't need to specify both parameters? That is not reflected in the docs.

File: centos_ci.sh [hunk 16]
```diff
@@ -292,14 +292,42 @@ function perform_release() {
...
+    git remote add custom-remote $REPO
```
Comment (hferentschik) [inline centos_ci.sh]: I guess these git commands work, but they don't really make sense to me. Why adding a remote if this is a once-off operation. You add the remote as shortcut if you use it a lot of times.

And why do you fetch and then track? IMO the most straight forward way here is what you see also documented in pull requests when you click on the "command line instructions" link

    # create the branch to host the code you want to build
    git checkout -b docs_branch master 
    # pull the changes into this branch
    git pull $REPO $BRANCH

So why so complicated?

File: centos_ci.sh [hunk 17]
```diff
@@ -292,14 +292,42 @@ function perform_release() {
...
+    git fetch custom-remote
+    git checkout -b remote-docs-branch --track custom-remote/$BRANCH
+  else
+    git checkout -b docs-branch --track origin/$BRANCH
+  fi
+
+  git log -n 1 # Check last commit
```
Comment (hferentschik) [inline centos_ci.sh]: "Check last commit" -> "Display last commit in log as reference"

You are not checking anything here

File: centos_ci.sh [hunk 18]
```diff
@@ -292,14 +292,42 @@ function perform_release() {
...
+  make gen_adoc_tar
+  build_openshift_origin_docs $(pwd)/docs/build/minishift-adoc.tar;
+
+  mkdir -p minishift/docs/ondemand/$REPO_OWNER-$BRANCH
+  cp -r openshift-docs/_preview/openshift-origin minishift/docs/ondemand/$REPO_OWNER-$BRANCH/ # Copy the openshift-origin
+  cp docs/build/minishift-adoc.tar minishift/docs/ondemand/$REPO_OWNER-$BRANCH/ # Copy doc tar
+  # http://stackoverflow.com/a/22908437/1120530; Using --relative as --rsync-path not working
+  RSYNC_PASSWORD=$RSYNC_PASSWORD rsync -av --delete --relative minishift/docs/ondemand/$REPO_OWNER-$BRANCH/ minishift@artifacts.ci.centos.org::minishift/
+  echo "Find docs tar here http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/$REPO_OWNER-$BRANCH/"
+  echo "Minishift docs is hosted at http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/$REPO_OWNER-$BRANCH/openshift-origin/latest/minishift/index.html"
```
Comment (hferentschik) [inline centos_ci.sh]: _are_

File: centos_ci.sh (hunk 16 above)
Comment (budhrg) [reply centos_ci.sh]: @hferentschik , I tested that here https://ci.centos.org/job/minishift-pr/2803/console and found that during pull it somethings ask to merge the commit and wait for user's input.

File: docs/source/contributing/writing-docs.adoc (hunk 13 above)
Comment (budhrg) [reply docs/source/contributing/writing-docs.adoc]: > I would think to do so, one needs to be logged in, but from what I recall from our discussions, we are not supposed to log into the instance. I cannot find the username/password either anyway. We have the access token, but that's a different beast, right? Something is not adding up here.

Yes, one need to log into Jenkins UI. `username/password` is mentioned in cdk mojo secret page.

> Something is not adding up here.

The piece which is missing here is user need to log into Jenkins UI to able to see the build with parameters option. Not sure how to tackle this in the documentation. 
Does following seems reasonable?

```
Currently, only maintainers have rights to trigger this job. If you want to stage your own version of the documentation, reach out to `#minishift` channel on https://freenode.net/[Freenode IRC] or send a request mail on https://lists.minishift.io/admin/lists/minishift.lists.minishift.io[mailing list].
```
 WDYT?

File: centos_ci.sh (hunk 16 above)
Comment (hferentschik) [reply centos_ci.sh]: `git pull -f`?

File: docs/source/contributing/writing-docs.adoc (hunk 13 above)
Comment (budhrg) [reply docs/source/contributing/writing-docs.adoc]: @hferentschik 

It is looking like

screenshot from 2017-11-29 15-11-17

Comment (hferentschik) [reply docs/source/contributing/writing-docs.adoc]: So why don't we not use URL based triggering as with the release. 
My issues is not that contributors cannot do it. I don't think we need to do anything special for them. My problem is that I can see the access token for triggering via URL documented in our internal docs, but not username/password.

File: centos_ci.sh [hunk 19]
```diff
@@ -292,14 +292,34 @@ function perform_release() {
   curl http://minibot.19cf262c.svc.dockerapp.io:9009/hubot/centosci -H "Content-Type: application/json" -d '{"payload":{"status":"success","message":'"\"$MESSAGE\""',"url":'"\"$URL\""'}}' || true
 }
 
+function perform_docs_publish() {
+  # REPO and BRANCH variables are populated via https://ci.centos.org/job/minishift-docs
+  REPO_OWNER=$(echo $REPO | cut -d"/" -f4)
+  cd gopath/src/github.com/minishift/minishift
+
+  git pull -f $REPO $BRANCH # Get changes from remote repo/branch
```
Comment (hferentschik) [inline centos_ci.sh]: No branch creation? Not needed? Did you test the 'pull -f' against a repo/branch which would not be a fast forward of the master?

File: docs/source/contributing/writing-docs.adoc [hunk 20]
```diff
@@ -159,3 +160,22 @@ $ rake build
 ----
 
 If the build completes successfully, the site is available under *_preview/openshift-origin/latest/welcome/index.html_*.
+
+[[building-openshift-docs-using-centosci]]
+=== Building the OpenShift Documentation using CentOS CI
+
+To get the view of {project} documentation integrated into the OpenShift documentation using CentOS CI, trigger the documentation build using following:
```
Comment (hferentschik) [inline docs/source/contributing/writing-docs.adoc]: "To get the view " -> "To view"

"to get the view" sounds awkward.

File: docs/source/contributing/writing-docs.adoc [hunk 21]
```diff
@@ -159,3 +160,22 @@ $ rake build
...
+
+----
+$ export API_KEY=<api-key>
+$ curl -H "$(curl --user minishift:$API_KEY 'https://ci.centos.org//crumbIssuer/api/xml?xpath=concat(//crumbRequestField,":",//crumb)')" -X POST https://ci.centos.org/job/minishift-docs/build --user "minishift:$API_KEY" --data-urlencode json='{"parameter": [{"name":"REPO", "value":"<user-repo>"}, {"name":"BRANCH", "value":"<docs-branch>"}]}'
```
Comment (hferentschik) [inline docs/source/contributing/writing-docs.adoc]: +1 :-)

What's about we export BRANCH and REPO as well and use it in the curl command. Then the curl command is cut&paste provided you have set the variables

**Comment by budhrg** [comment]
> Eventually, I would still like to rework the whole script and split it up. For now we should be good.

Create an issue https://github.com/minishift/minishift/issues/1763.

**Comment by hferentschik** [comment]
Merged. Thanks.
